<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing ancient DNA data with the R package tfa • tfa</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing ancient DNA data with the R package tfa">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tfa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tfa-vignette.html">Analyzing ancient DNA data with the R package tfa</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Analyzing ancient DNA data with the R package tfa</h1>
                        <h4 class="author">Olivier François - Flora Jay</h4>
            
            <h4 class="date">2020-03-31</h4>
      
      
      <div class="hidden name"><code>tfa-vignette.Rmd</code></div>

    </div>

    
    
<p><strong>Summary:</strong> The R package <code>tfa</code> implements a factor analysis (FA) algorithm for temporal DNA or ancient DNA (aDNA) samples, ajusting individual scores for the effect of allele frequency drift through time. Based on the adjusted factors, the program can estimate ancestry proportions for a target population or a subset of target individuals given specified source populations. This vignette illustrates the use of <code>tfa</code> with samples of Eurasian aDNA extracted from <a href="https://reich.hms.harvard.edu/">David Reich’s lab</a> database.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tfa)</code></pre></div>
<div id="principle-of-the-method" class="section level4">
<h4 class="hasAnchor">
<a href="#principle-of-the-method" class="anchor"></a>Principle of the method</h4>
<p>The FA model uses a decomposition of the data matrix into a product of factor and loading matrices plus correction factors and residual noise. The correction factors represented the main improvement over classical FA or principal component analysis (PCA). In <code>tfa</code> the correction factors are related to the covariance function of the stochastic drift process. The default model for the covariance function is the Brownian model. This default corresponds to the diffusion approximation of allele frequency drift in a random mating population conditional on the non-fixation of alleles in the population. The approach is flexible and allows expert users to customize their own model and implement other covariance functions. The correction factors are obtained after a spectral decomposition of the covariance matrix, and the corresponding effect size estimates are obtained in a Bayesian framework. Explicit solutions are obtained for the effect size estimates and corrected factors. The hyperparameters for the FA model are the drift parameter and the number of latent factors.</p>
</div>
<div id="factor-analysis-of-adna-samples" class="section level4">
<h4 class="hasAnchor">
<a href="#factor-analysis-of-adna-samples" class="anchor"></a>Factor analysis of aDNA samples</h4>
<p>This section will illustrate the use of the <code>tfa</code> function to display a geometric representation of 118 aDNA samples from prehistoric Eurasia. This example data set was extracted from a database available from <a href="https://reich.hms.harvard.edu/">David Reich’s lab</a> considering individuals from Neolithic and Bronze Age Great Britain, Steppe (Yamnaya), Anatolia (Early Farmers), and from Serbia (Western Hunter-Gatherers).</p>
<p>Ten thousands SNP genotypes were filtered for high-quality, relatively high coverage and low number of missing values. Individual metadata and sample ages are provided as separate objects. The command below shows how to load the data in the R environment and which groups have been included in the data set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(england_ba)
<span class="co"># Ancient DNA from Bronze Age Great Britain samples</span>
<span class="co"># including Yamnaya, early farmers (Anatolia) and hunter-gatherers (Serbia)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/attach.html">attach</a></span>(England_BA)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(meta<span class="op">$</span>Group.ID), <span class="dt">las =</span> <span class="dv">3</span>, <span class="dt">cex.axis =</span> .<span class="dv">5</span>, <span class="dt">ylab =</span> <span class="st">"Count"</span>, <span class="dt">lwd =</span> <span class="dv">5</span>, <span class="dt">col =</span> <span class="st">"blue"</span>)</code></pre></div>
<p><img src="tfa-vignette_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>Prehistoric dates for each sample are provided in years cal BP.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(age, <span class="dt">col =</span> <span class="st">"lightblue"</span>, <span class="dt">xlab =</span> <span class="st">"Age cal BP"</span>)</code></pre></div>
<p><img src="tfa-vignette_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>The data set contains no missing genotypes. The missing values were imputed from a much larger database. This was done in a preliminary analysis by using a completion algorithm implemented in the R package <code>LEA</code>. Running a standard PCA on the data matrix clearly shows there are three main axes of variation in the genotypic data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  pca &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(genotype)
  <span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(pca, <span class="dt">col =</span> <span class="st">"palegreen"</span>)</code></pre></div>
<p><img src="tfa-vignette_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>A first step of analysis consists of adjusting the genotypic data for coverage. For those particular filtered genotypes, the impact of coverage bias on analyses is not substantial. In general, coverage may strongly bias the results of factor analyses, and this important to consider adjusting for it. To adjust for coverage, the <code>tfa</code> package contains the function <code><a href="../reference/coverage_adjust.html">coverage_adjust()</a></code> which implements a correction based on a latent factor regression model. According to the PCA result, three factors will be used in the latent factor regression model.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">coverage &lt;-<span class="st"> </span>meta<span class="op">$</span>Coverage
geno &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coverage_adjust.html">coverage_adjust</a></span>(genotype, coverage, <span class="dt">K =</span> <span class="dv">3</span>, <span class="dt">log =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>For running FA of the corrected data matrix (<code>geno</code>), the number of factors (<code>k</code>), and the drift parameter (<code>lambda</code>) must be specified. The <strong>number of factors</strong> could be determined by PCA or other methods for analysis of population structure. It could be taken equal to the number of axes in the PCA of the <code>geno</code> matrix. Since the <code><a href="../reference/tfa.html">tfa()</a></code> approach is based a low-rank approximation of the data matrix, a lower value such as <code>k = 2</code> could be accepted. The value <code>k = 2</code> could be useful to explore relationships of samples when two ancestral populations are specified. The <strong>drift parameter</strong> (<code>lambda</code>) can be determined by a grid search procedure. Here, this parameter was chosen as the value for which the effect of age was removed from the <strong>third</strong> factor of the <code>tfa</code> analysis performed below, which was repeated several times with different values of <code>lambda</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  mod  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/tfa.html">tfa</a></span>(<span class="dt">Y =</span> geno,
              <span class="dt">sample_ages =</span> age,
              <span class="dt">k =</span> <span class="dv">3</span>,
              <span class="dt">lambda =</span> <span class="fl">5e-1</span>)</code></pre></div>
<p>The object <code>mod</code> contains <code>k</code> factors adjusted for temporal drift in allele frequency. The results are contained in argument <code>mod$u</code>. The argument <code>mod$sing</code> contains the singular value (variance) associated with each factor.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> var_explained &lt;-<span class="st"> </span>mod<span class="op">$</span>singular.values[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]<span class="op">/</span>(<span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(mod<span class="op">$</span>singular.values))
 var_explained
<span class="co">#&gt; [1] 0.4274355 0.3227561 0.2498085</span></code></pre></div>
<p>The object <code>mod$u</code> can be used to visualize the samples in their first axes. The result looks like a PC plot showing the relative positions of Great Britain samples with respect to their putative sources of ancestry.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(mod<span class="op">$</span>u, <span class="dt">pch =</span> <span class="dv">19</span>, <span class="dt">cex =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">"grey90"</span>, <span class="dt">xlab =</span> <span class="st">"Factor 1"</span>, <span class="dt">ylab =</span> <span class="st">"Factor 2"</span>)

  center_yamnaya &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(mod<span class="op">$</span>u[meta<span class="op">$</span>Group.ID <span class="op">==</span><span class="st"> "Russia_Yamnaya"</span>,], <span class="dv">2</span>, mean)
  center_anatolia &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(mod<span class="op">$</span>u[meta<span class="op">$</span>Group.ID <span class="op">==</span><span class="st"> "Anatolia_N"</span>,], <span class="dv">2</span>, mean) 
  center_hg &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(mod<span class="op">$</span>u[meta<span class="op">$</span>Group.ID <span class="op">==</span><span class="st"> "Serbia_HG"</span>,], <span class="dv">2</span>, mean)

  <span class="kw"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(center_yamnaya, center_anatolia, center_hg, center_yamnaya))
  
  <span class="kw"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(mod<span class="op">$</span>u[meta<span class="op">$</span>Group.ID <span class="op">==</span><span class="st"> "Russia_Yamnaya"</span>,], <span class="dt">pch =</span> <span class="dv">8</span>, <span class="dt">cex =</span> .<span class="dv">6</span>, <span class="dt">col =</span> <span class="st">"darkblue"</span>)
  <span class="kw"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(mod<span class="op">$</span>u[meta<span class="op">$</span>Group.ID <span class="op">==</span><span class="st"> "Anatolia_N"</span>,], <span class="dt">pch =</span> <span class="dv">8</span>, <span class="dt">cex =</span> .<span class="dv">6</span>, <span class="dt">col =</span> <span class="st">"salmon3"</span>)
  <span class="kw"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(mod<span class="op">$</span>u[meta<span class="op">$</span>Group.ID <span class="op">==</span><span class="st">  "Serbia_HG"</span>,], <span class="dt">pch =</span> <span class="dv">8</span>, <span class="dt">cex =</span> .<span class="dv">6</span>, <span class="dt">col =</span> <span class="st">"olivedrab"</span>)
  <span class="kw"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(mod<span class="op">$</span>u[meta<span class="op">$</span>Group.ID <span class="op">==</span><span class="st"> "England_Bell_Beaker"</span>,], <span class="dt">pch =</span> <span class="dv">19</span>, <span class="dt">cex =</span> .<span class="dv">6</span>, <span class="dt">col =</span> <span class="st">"yellow4"</span>)
  <span class="kw"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(mod<span class="op">$</span>u[meta<span class="op">$</span>Group.ID <span class="op">==</span><span class="st"> "England_BA"</span>,], <span class="dt">pch =</span> <span class="dv">19</span>, <span class="dt">cex =</span> .<span class="dv">6</span>, <span class="dt">col =</span> <span class="st">"yellow3"</span>)
  <span class="kw"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(mod<span class="op">$</span>u[meta<span class="op">$</span>Group.ID <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"England_N"</span>, <span class="st">"Scotland_N"</span>),], <span class="dt">pch =</span> <span class="dv">19</span>, <span class="dt">cex =</span> .<span class="dv">6</span>, <span class="dt">col =</span> <span class="st">"salmon1"</span>)

  <span class="kw"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(<span class="dt">x =</span> <span class="dv">20</span>, <span class="dt">y =</span> <span class="op">-</span><span class="dv">15</span>, <span class="dt">cex =</span> .<span class="dv">6</span>,
         <span class="dt">legend =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Early Farmers"</span>, <span class="st">"Hunter Gatherers"</span>, <span class="st">"Steppe"</span>), 
         <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"salmon3"</span>, <span class="st">"olivedrab"</span>, <span class="st">"darkblue"</span>), <span class="dt">pch =</span> <span class="dv">8</span>)
  <span class="kw"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(<span class="dt">x =</span> <span class="dv">10</span>, <span class="dt">y =</span> <span class="op">-</span><span class="dv">15</span>, <span class="dt">cex =</span> .<span class="dv">6</span>,
         <span class="dt">legend =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Neolithic GB"</span>, <span class="st">"Bronze Age GB"</span>, <span class="st">"Bell Beaker"</span>), 
         <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"salmon1"</span>, <span class="st">"yellow3"</span>, <span class="st">"yellow4"</span>), <span class="dt">pch =</span> <span class="dv">19</span>)</code></pre></div>
<p><img src="tfa-vignette_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div id="ancestry-coefficients" class="section level4">
<h4 class="hasAnchor">
<a href="#ancestry-coefficients" class="anchor"></a>Ancestry coefficients</h4>
<p>In addition to geometric representations, the R package <code>tfa</code> can provide estimates of ancestry proportions for a target population or for a list of target individuals given specified source population samples. The number of sources must be less or equal than the number of factors plus one. For this analysis, the metadata object must contain a column named <strong>Instance.ID</strong> indicating the ID of each individual and another column named <strong>Group.ID</strong> indicating the group ID of each individual in the study.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(meta)
<span class="co">#&gt; [1] "Instance.ID" "Group.ID"    "Country"     "Coverage"</span></code></pre></div>
<p>Since the data.frame <code>meta</code> contains the correct information, it can be converted into an object of class <code>tfa_metadata</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metadata &lt;-<span class="st"> </span><span class="kw"><a href="../reference/as.tfa_metadata.html">as.tfa_metadata</a></span>(meta)</code></pre></div>
<p>Now consider <code>England_Bell_Beaker</code> and <code>England_BA</code> as target populations and <code>Anatolia_N</code> (Early Farmers), <code>Russia_Yamnaya</code> (Steppe Pastoralists), <code>Serbia_HG</code> (Western Hunter-Gatherers) as potential source populations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  target =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"England_Bell_Beaker"</span>, <span class="st">"England_BA"</span>)
  source =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Anatolia_N"</span>, <span class="st">"Russia_Yamnaya"</span>, <span class="st">"Serbia_HG"</span>)</code></pre></div>
<p>The function <code>ancestry_coefficients</code> uses basic linear algebra to compute the coordinates of each target population in the system formed by the centers of source populations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw"><a href="../reference/ancestry_coefficients.html">ancestry_coefficients</a></span>(<span class="dt">model =</span> mod,
                        <span class="dt">metadata =</span> metadata,
                        <span class="dt">source =</span> source,
                        <span class="dt">target =</span> target)
<span class="co">#&gt;                     Anatolia_N Russia_Yamnaya  Serbia_HG</span>
<span class="co">#&gt; England_Bell_Beaker  0.2673478      0.6309998 0.10165241</span>
<span class="co">#&gt; England_BA           0.3283689      0.5723816 0.09924954</span></code></pre></div>
<p>The Bell Beaker samples share a large fraction of their ancestry with steppe pastoralists. Changing the Anatolian population for another population of Early Farmers (<code>England_N</code>) can confirm this result.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw"><a href="../reference/ancestry_coefficients.html">ancestry_coefficients</a></span>(<span class="dt">model =</span> mod,
                        <span class="dt">metadata =</span> metadata,
                        <span class="dt">source =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"England_N"</span>, <span class="st">"Russia_Yamnaya"</span>, <span class="st">"Serbia_HG"</span>),
                        <span class="dt">target =</span> target)
<span class="co">#&gt;                     England_N Russia_Yamnaya  Serbia_HG</span>
<span class="co">#&gt; England_Bell_Beaker 0.2844837      0.6732371 0.04227924</span>
<span class="co">#&gt; England_BA          0.3494159      0.6242594 0.02632469</span></code></pre></div>
<p>The result shows that the amount of ancestry shared with the hunter-gatherers reduces when farmers are sampled from a closer geographic source. This suggests that the hunter-gatherer contribution to the England Neolithic gene pool may be substantial. To check this hypothesis, consider <code>England_N</code> as the target with the set of sources considered at the start.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw"><a href="../reference/ancestry_coefficients.html">ancestry_coefficients</a></span>(<span class="dt">model =</span> mod,
                        <span class="dt">metadata =</span> metadata,
                        <span class="dt">source =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Anatolia_N"</span>, <span class="st">"Russia_Yamnaya"</span>, <span class="st">"Serbia_HG"</span>),
                        <span class="dt">target =</span> <span class="st">"England_N"</span>)
<span class="co">#&gt;           Anatolia_N Russia_Yamnaya Serbia_HG</span>
<span class="co">#&gt; England_N  0.9397651     -0.1484701  0.208705</span></code></pre></div>
<p>A negative value for the <code>Russia_Yamnaya</code> source suggests that this source did not contribute to the gene pool of Neolithic England samples, and that it should be removed from the set of sources.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw"><a href="../reference/ancestry_coefficients.html">ancestry_coefficients</a></span>(<span class="dt">model =</span> mod,
                        <span class="dt">metadata =</span> metadata,
                        <span class="dt">source =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Anatolia_N"</span>, <span class="st">"Serbia_HG"</span>),
                        <span class="dt">target =</span> <span class="st">"England_N"</span>)
<span class="co">#&gt;           Anatolia_N Serbia_HG</span>
<span class="co">#&gt; England_N  0.8480804 0.1519196</span></code></pre></div>
<p>After doing this, the result indeed suggests a substantial contribution of hunter-gatherers to Neolithic England farmers. The analysis can be summarized in a table of ancestry coefficients as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ancestry.B &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ancestry_coefficients.html">ancestry_coefficients</a></span>(<span class="dt">model =</span> mod,
                                  <span class="dt">metadata =</span> metadata,
                                  <span class="dt">source =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Serbia_HG"</span>, <span class="st">"Anatolia_N"</span>, <span class="st">"Russia_Yamnaya"</span>),
                                  <span class="dt">target =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"England_Bell_Beaker"</span>, <span class="st">"England_BA"</span>))

ancestry.N &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ancestry_coefficients.html">ancestry_coefficients</a></span>(<span class="dt">model =</span> mod,
                                  <span class="dt">metadata =</span> metadata,
                                  <span class="dt">source =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Serbia_HG"</span>, <span class="st">"Anatolia_N"</span>),
                                  <span class="dt">target =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"England_N"</span>, <span class="st">"Scotland_N"</span>))

ancestry.N &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(ancestry.N, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>,<span class="dv">0</span>))
ancestry &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(ancestry.B, ancestry.N)
ancestry
<span class="co">#&gt;                      Serbia_HG Anatolia_N Russia_Yamnaya</span>
<span class="co">#&gt; England_Bell_Beaker 0.10165241  0.2673478      0.6309998</span>
<span class="co">#&gt; England_BA          0.09924954  0.3283689      0.5723816</span>
<span class="co">#&gt; England_N           0.15191961  0.8480804      0.0000000</span>
<span class="co">#&gt; Scotland_N          0.17030362  0.8296964      0.0000000</span></code></pre></div>
<p>The results can be represented by a barplot as follows. The green, salmon and blue colors correspond to hunter-gatherers, early farmers and steppe pastoralists respectively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(ancestry), 
          <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"olivedrab"</span>, <span class="st">"salmon3"</span>, <span class="st">"darkblue"</span>), 
          <span class="dt">main =</span> <span class="st">"Ancestry coefficients"</span>)</code></pre></div>
<p><img src="tfa-vignette_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
<div id="individual-ancestry-coefficients" class="section level4">
<h4 class="hasAnchor">
<a href="#individual-ancestry-coefficients" class="anchor"></a>Individual ancestry coefficients</h4>
<p>Since FA is an individual-based method, a geometric approach can be used to compute ancestry coefficients for each individual, and to display results looking like a popular STRUCTURE barplot. For Bronze Age and Bell Beaker samples, this can be done as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  source =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Anatolia_N"</span>, <span class="st">"Russia_Yamnaya"</span>, <span class="st">"Serbia_HG"</span>)
  target =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"England_Bell_Beaker"</span>, <span class="st">"England_BA"</span>)

  Q_matrix &lt;-<span class="st">  </span><span class="kw"><a href="../reference/ancestry_coefficients.html">ancestry_coefficients</a></span>(<span class="dt">model =</span> mod,
                                     <span class="dt">metadata =</span> metadata,
                                     <span class="dt">source =</span> source,
                                     <span class="dt">target =</span> target,
                                     <span class="dt">individual =</span> <span class="ot">TRUE</span>)
  
  <span class="kw"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(Q_matrix), 
          <span class="dt">border =</span> <span class="ot">NA</span>, 
          <span class="dt">space =</span> <span class="dv">0</span>, 
          <span class="dt">axisnames =</span> <span class="ot">FALSE</span>, 
          <span class="dt">col =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"salmon3"</span>, <span class="st">"darkblue"</span>, <span class="st">"olivedrab"</span>),
          <span class="dt">legend.text =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Anatolia"</span>, <span class="st">"Steppe"</span>, <span class="st">"WHG"</span>),
          <span class="dt">args.legend =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">y =</span> <span class="fl">0.8</span>, <span class="dt">cex =</span> .<span class="dv">6</span>),
          <span class="dt">ylab =</span> <span class="st">"Ancestry proportions"</span>)
  
  <span class="kw"><a href="https://rdrr.io/r/graphics/axis.html">axis</a></span>(<span class="dv">1</span>, <span class="dt">at =</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(Q_matrix), <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(Q_matrix), <span class="dt">las =</span> <span class="dv">3</span>, <span class="dt">cex.axis =</span> .<span class="dv">4</span>)</code></pre></div>
<p><img src="tfa-vignette_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="kw"><a href="https://rdrr.io/r/base/detach.html">detach</a></span>(England_BA)
  
<span class="co">#clear environment   </span>
<span class="co">#rm(list = ls())</span></code></pre></div>
<p>The levels of ancestry are consistent across the samples from Bell Beaker and Bronze Age samples.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Olivier Francois.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
